### PR TITLE
Implementing Sanity-Check

### LINKED ISSUE #14: [Feature Request] Implementation of Sanity Check
## Implementation of Sanity Check

The implementation can be divided into the following parts:

### SanityCheckResult
This class should implement the [SerializableResult](https://saliency-metrics.readthedocs.io/en/latest/metrics.html#saliency_metrics.metrics.SerializableResult). It aims to record the sanity check result of each image, and can be dumped into a JSON file.

The class should look like this
```python3
class SanityCheckResult(SerializableResult):
    def __init__(self, summarized: bool = False, ...) -> None:
        ....

    def add_single_result(self,  single_result: Dict) -> None:
        ...

    def dump(self, file_path: str) -> None:
        ...
```
An `SanityCheckResult` instance caches the results of images in a container: e.g. `dict` or `list`. The result of a single image should be a dict that contains following fields:
* `"img_path"`: the path of the input image.
* `"ssim"`: `ndarray` with shape `(num_perturb_layers, )`. It is the array of structural similarity associated with various layer perturbation settings.

The `dump` method dumps the cached result into a JSON file. [mmcv.dump](https://github.com/open-mmlab/mmcv/blob/c70fafebf44b99422e75f35eee76cf28618f1e2a/mmcv/fileio/io.py#L69) can be used as a helper function here. 

The flag `summarized` specifies whether we dump the raw result or we compute the means and stds of SSIM associated with different layer perturbation settings. I.e., the raw result should be like this:
```JSON
{
    {"img_path": "path/to/img_0.jpg", "ssim": [0.9, 0.8, 0.7]},
    {"img_path": "path/to/img_1.jpg", "ssim": [0.8, 0.7, 0.6]},
}
```
while the summarized result should be like this:

```JSON
{
    "mean_ssim": [0.85, 0.75, 0.65],
    "std_ssim": [0.031, 0.032, 0.026],
    "num_samples": 2,
}
```

## SanityCheck Metric
The Class `SanityCheck` should implement the protocol [ReInferenceMetric](https://saliency-metrics.readthedocs.io/en/latest/metrics.html#saliency_metrics.metrics.ReInferenceMetric). It progressively perturbs the model layers, and compare the saliency map obtained under each perturbation setting with the original saliency map, i.e., the saliency map obtained from unperturbed model. The similarity saliency maps are measured by [structural similarity](https://github.com/scikit-image/scikit-image/blob/main/skimage/metrics/_structural_similarity.py#L15-L254).

The class should look like this:

```python3
class SanityCheck(ReInferenceMetric):
    def __init__(self, classifier: nn.Module, perturb_layers: List[str], attr_method: AttributionMethod, ssim_args: Optional[Dict] = None, summarized: bool = False) -> None:
        self._result = SanityCheckResult(summarized)
        self._ori_state_dict = deepcopy(classifier.state_dict())
        self._classifier_layers: List[str] = self._filter_names([n[0] for n in classifier])
        ...

    def _reload_ckpt(self, device: Optional[Union[str, torch.device]] = None) -> None:
        ...
    
    def evaluate(self, torch.Tensor, smap: np.ndarray, target: int) -> Dict:
        ...
    
    def update(self, single_result: Dict) -> None:
        ...
    
    def _sanity_check_single(self, img: Tensor, target: int, consecutive_perturb_layers: List[str]) -> float:

    def perturb_classifier(self, classifier: nn.Module, layers: List[str]) -> None:
        ...

    @staticmethod
    def _filter_names(names: List[str]) -> List[str]:
        res: List[str] = []
        for i in range(len(names) - 1):
            if not names[i] in names[i+1]:
                res.append(names[i])
        res.append(names[-1])
        return res
```

1. In the initialization method, `attr_method` represents the attribution method that implements [AttributionMethod](https://github.com/scikit-image/scikit-image/blob/main/skimage/metrics/_structural_similarity.py#L15-L254). `_ori_state_dict` is the state dict of the unperturbed model. 

2. The argument `perturb_layers` stores the progressive perturbation settings. Each element is a layer name, and it can be interpreted as "perturbing from the last layer to this layer". 

3. The static method `_filter_names` helps to filter the layer names. The layer names contained a recusive named module can be like this:
```python
['', 'conv1', 'bn1', 'act1', 'maxpool', 'layer1', 'layer1.0', 'layer1.0.conv1', 'layer1.0.bn1', 'layer1.0.act1', 'layer1.0.conv2', 'layer1.0.bn2', 'layer1.0.act2', 'layer1.1', 'layer1.1.conv1', 'layer1.1.bn1', 'layer1.1.act1', 'layer1.1.conv2', 'layer1.1.bn2', 'layer1.1.act2', 'layer2', 'layer2.0', 'layer2.0.conv1', 'layer2.0.bn1', 'layer2.0.act1', 'layer2.0.conv2', 'layer2.0.bn2', 'layer2.0.act2', 'layer2.0.downsample', ...]
```
After being filtered, they are like this:
```python
["conv1", "bn1", "act1", "maxpool", "layer1", "layer2", ...]
```

4. The `_reload_ckpt` function does the following jobs:
* reload the original (unperturbed) state_dict
* send the model to the device if it is not None
* turn the classifier into eval mode.
* freeze the entire model.

It should be called when performing sanity check under each perturbation setting, i.e. at the beginning in the for-loop:
```python3
for layer in self.perturb_layers:
    self.reload_ckpt(device=img.device)
    ...
    self._sanity_check_single(...)

```

5. `_sanity_check_single` performs sanity check under a specific perturbation settings, which is specified by the argument `consecutive_perturb_layers`.

6. `perturb_classifier` perturb a trunk of consecutive model layers.

### PR DESCRIPTION
**Pull request by saran9991**
## Short Description

Fixes #14 .

## Long Description
Implementation of the classes and functions

## Checklist

- [ x] Code linting are passed. (Please run `pre-commit run --all-files`).
- [ ] Mypy checks are passed. (Please run `mypy saliency_metrics`).
- [ ] Unit tests are added for new code.
- [ ] Documentation is added for the new code.
- [ ] Unit tests are passed. (Please run `sh tests/run_tests.sh`).
- [ ] Docs are successfully built. (Please run `cd docs/ && make html` and host the webpage locally).

### CODE REVIEW

File: saliency_metrics/metrics/sanity_check/sanity_check_result.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    def __init__(self, summarized=True) -> None:
+        self.summarized = summarized
+        self.results: List = []
+
+    def add_single_result(self, single_result: Dict) -> None:
+        self.results.append(single_result)
+
+    def dump(self, file_path: str) -> None:
+        if self.summarized:
+            ssim_list: List = []
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check_result.py:20]: line 20 - 23 can be more compact:

```python

ssim_array = np.asarray([r["ssim"] for r in self.results])
```

File: saliency_metrics/metrics/sanity_check/sanity_check_result.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            for result in self.results:
+                ssim_list.append(result["ssim"])
+            ssim_array = np.stack(ssim_list)
+            summarized_result = {
+                "mean_ssim": np.mean(ssim_array, axis=0),
+                "std_ssim": np.std(ssim_array, axis=0),
+                "num_samples": len(ssim_list),
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check_result.py:27]: If the line 20 - 23 is changed as suggested, here you can write `len(ssim_array)`.

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+from copy import deepcopy
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+import torch
+from skimage.metrics import structural_similarity
+from torch import Tensor
+
+from saliency_metrics.metrics.build_metric import METRICS, ReInferenceMetric
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:9]: line 9 - 12: Please use relative path to shorten the imports. E.g.,
`from ..build_metric import METRICS, ReInferenceMetric`

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+from saliency_metrics.metrics.sanity_check.sanity_check_result import SanityCheckResult
+from saliency_metrics.metrics.serializable_result import SerializableResult
+from saliency_metrics.models.model_utils import freeze_module
+
+
+class SanityCheck(ReInferenceMetric):
+    def __init__(self, ori_model: torch.nn.Module, perturb_layers: List[str], summarized: bool = True) -> None:
+        self._result = SanityCheckResult(summarized)
+        self.ori_model = ori_model  # Original Model
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:18]: Please move the `# Original model` to a new line above. The same for line 20, 41, and 47.

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        self._ori_state_dict = self.ori_model.state_dict()
+        self.model = deepcopy(self.ori_model)  # Model which will be used for performing randomization on
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:20]: rename to `_ori_state_dict`. We don't need to perform `memcpy` to copy all the contents of the model, instead, we just need the state dict.

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        self.sd = self.model.state_dict()
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:21]: Why store another copy? (Please do not use such ambiguous name like `sd`.)

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        self.perturb_layers = perturb_layers
+        self.all_layers: List = []
+        for i, x in self.ori_model.named_parameters():
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:24]: Why named parameters? Named parameter returns a tuple of `(name, parameter)`, and one layer can contain multiple parameters. Please CHECK THE GUIDELINE WHICH I SUGGESTED IN THE GITHUB ISSUE.

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            self.all_layers.append(i)
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:25]: Please do dot make the name so ambiguous. the first element is `name`, why use `i`, which often stands for index.

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        self._classifier_layers = self._filter_names(self.all_layers)
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:26]: `self._classifier_layers = self._filter_names([n[0] for n in classifier.named_modules()])`

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        smap1 = np.random.randint(0, 255, (10, 10), dtype=np.uint8)  # Custom SMAP and SSIM
+        return structural_similarity(smap1, smap, data_range=255)
+
+    def _perturb_classifier(self, model: torch.nn.Module, layers: List[str]) -> None:
+        for i in layers:
+            prng = np.random.RandomState(42)
+            new_parameters = prng.uniform(0, 0.2, self.sd[i].shape)  # Random Numbers Sampled from Uniform Distribution
+            self.sd[i] = torch.tensor(new_parameters)
+
+    def _reload_ckpt(self, device: Optional[Union[str, torch.device]] = "cpu") -> None:
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:50]: `Optional` is used here. Then why the default value is a string? It should be `None`.

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+from torch import Tensor
+
+from saliency_metrics.metrics.build_metric import METRICS, ReInferenceMetric
+from saliency_metrics.metrics.sanity_check.sanity_check_result import SanityCheckResult
+from saliency_metrics.metrics.serializable_result import SerializableResult
+from saliency_metrics.models.model_utils import freeze_module
+
+
+class SanityCheck(ReInferenceMetric):
+    def __init__(self, ori_model: torch.nn.Module, perturb_layers: List[str], summarized: bool = True) -> None:
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:16]: Please stick to the API as I suggested in the GitHub Issue. `ori_model` should be `classifier`.

Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:16]: Where is the `attr_method`???

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        return dict(ssim=ssim_list)
+
+    def _sanity_check_single(self, smap: np.ndarray, target: int, consecutive_perturb_layers: List[str]) -> float:
+        self._perturb_classifier(self.model, consecutive_perturb_layers)
+        smap1 = np.random.randint(0, 255, (10, 10), dtype=np.uint8)  # Custom SMAP and SSIM
+        return structural_similarity(smap1, smap, data_range=255)
+
+    def _perturb_classifier(self, model: torch.nn.Module, layers: List[str]) -> None:
+        for i in layers:
+            prng = np.random.RandomState(42)
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:46]: Do not call numpy rand. PyTorch has its own functionalities for creating random tensors or re-filling a tensor with random numbers.

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+            new_parameters = prng.uniform(0, 0.2, self.sd[i].shape)  # Random Numbers Sampled from Uniform Distribution
+            self.sd[i] = torch.tensor(new_parameters)
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:48]: Why rewrite the state dict stored in `Self`? The `model` is passed in this function but it is never used.

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+            consecutive_perturb_layers: List = []
+            for j in self._classifier_layers[self._classifier_layers.index(layer) :]:
+                consecutive_perturb_layers.append(j)
+            ssim_float = self._sanity_check_single(smap, target, consecutive_perturb_layers)
+            ssim_list.append(ssim_float)
+        return dict(ssim=ssim_list)
+
+    def _sanity_check_single(self, smap: np.ndarray, target: int, consecutive_perturb_layers: List[str]) -> float:
+        self._perturb_classifier(self.model, consecutive_perturb_layers)
+        smap1 = np.random.randint(0, 255, (10, 10), dtype=np.uint8)  # Custom SMAP and SSIM
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:41]: Why random smap? The new smap should be obtained by perform attribution on the perturbed model.

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+        for i, x in self.ori_model.named_parameters():
+            self.all_layers.append(i)
+        self._classifier_layers = self._filter_names(self.all_layers)
+
+    def evaluate(self, image: Tensor, smap: np.ndarray, target: int, **kwargs) -> Dict:
+        ssim_list: List = []
+        for layer in self.perturb_layers:
+            self._reload_ckpt()
+            consecutive_perturb_layers: List = []
+            for j in self._classifier_layers[self._classifier_layers.index(layer) :]:
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:33]: Wrong. 1. The layers should be perturbed from the last to the first one. But `index` only gives you the index of the first seen `layer`. However, it is possible that a model has two layers with identical names. In this case, the implementation is wrong. 2. PLEASE DO NOT USE `j` in this case, because `j` is usually used to denote index. But here the element from the list should be a string. This is very ambiguous and will dramatically increase maintenance burden.

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+                consecutive_perturb_layers.append(j)
+            ssim_float = self._sanity_check_single(smap, target, consecutive_perturb_layers)
+            ssim_list.append(ssim_float)
+        return dict(ssim=ssim_list)
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:37]: `**kwargs` might contain more information about the input image.
```python3
img_path = kwargs.get("img_path", "")
return dict(img_path=img_path, ssim=ssim_list)
```

File: saliency_metrics/metrics/sanity_check/sanity_check.py [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+        self._ori_state_dict = self.ori_model.state_dict()
+        self.model = deepcopy(self.ori_model)  # Model which will be used for performing randomization on
+        self.sd = self.model.state_dict()
+        self.perturb_layers = perturb_layers
+        self.all_layers: List = []
+        for i, x in self.ori_model.named_parameters():
+            self.all_layers.append(i)
+        self._classifier_layers = self._filter_names(self.all_layers)
+
+    def evaluate(self, image: Tensor, smap: np.ndarray, target: int, **kwargs) -> Dict:
```
Comment (sandylaker) [inline saliency_metrics/metrics/sanity_check/sanity_check.py:28]: use `img` instead of `image` in order to be consistent with other parts of this library.
